rfd/0073-kube-secret-storage.md

```diff
@@ -0,0 +1,501 @@
+
+```yaml
+- apiGroups: [""]
+  # objects is "secrets"
+  resources: ["secrets"]
+  verbs: ["create", "get", "update","watch", "list"]
+```
+
+This allows the Kube Agent to read, update and list secrets using the credentials that Kubernetes mounts for each pod running in the cluster.
+
+### Teleport Changes
+
+#### Kube Secret as storage backend
+
+If secret storage is enabled, the Teleport Kube agent initializes with Kubernetes secret [backend storage](https://goteleport.com/docs/setup/reference/backends/). The backend storage availability for Teleport will be the following:
+
+| Data type | Description | Supported storage backends |
+|---|---|---|
+| core cluster state | Cluster configuration (e.g. users, roles, auth connectors) and identity (e.g. certificate authorities, registered nodes, trusted clusters). | Local directory (SQLite), etcd, AWS DynamoDB, GCP Firestore, self-hosted PostgreSQL/CockroachDB (Preview) |
+| audit events | JSON-encoded events from the audit log (e.g. user logins, RBAC changes) | Local directory, AWS DynamoDB, GCP Firestore |
+| session recordings | Raw terminal recordings of interactive user sessions | Local directory, AWS S3 (and any S3-compatible product), GCP Cloud Storage |
+| teleport instance state | ID and credentials of a non-auth teleport instance (e.g. node, proxy, kube) | Local directory, Kube Secret (only available for role=kube) |
+
+The storage backend will be responsible for managing the Kubernetes secret, i.e. reading and updating its contents, in order to create a transparent storage backend.
+
+The configuration of the Kubernetes' Secret storage backend is done by adding the `storage` section to the teleport config.
+
+```yaml
+teleport:
+...
+  storage:
+    # Type of backend: either "local" or "kubernetes_secret", by default "local"
+    type: kubernetes_secret
+...
+```
+
+If the identity secret exists in Kubernetes and has the node identity on it (entry for `$TELEPORT_REPLICA_NAME`), the storage engine will parse and return the keys to the Agent so it can use them to authenticate in the Teleport Cluster. If the cluster access operation is successful, the agent will be available for usage, but if the access operation fails because the Teleport Auth does not validate the node credentials, the Agent will log an error providing insightful information about the failure cause.
+
+If the identity secret does not exist or is empty, the Kube Agent will try to join the cluster with the invite token provided. If the invite token is valid (has details in the Teleport Cluster and did not expire yet), Teleport Cluster will reply with the agent identity. Given the identity, the Kube Agent will issue a write request to the storage backend that results in the update of the content in the secret `{{ .Release.Name }}-identity`.
+
+If the invite token is not valid or has expired, the Agent could not join the cluster and it will stop and log a meaningful error message.
```

Comment:
Can you also add a section about handling CA rotation? All agents' identities are rotated during CA rotation as well.

If we just transparently replace identity's read/write with Kube Secrets when running on Kube, it should just be handled automatically but might be worth mentioning.